File: frontend/src/services/apiClient.js
/**
 * Central API Client (Security-Hardened)
 * =======================================
 * Handles ALL backend communication.
 * 
 * Token Strategy:
 *   - Access token stored in memory (not localStorage) to prevent XSS theft
 *   - Refresh token stored in HttpOnly cookie (set by backend)
 *   - On 401, automatically attempts token refresh before redirecting
 *   - Falls back to session cookies for desktop browsers
 * 
 * Migration note: Also checks localStorage for backward compatibility
 * with existing sessions, but new tokens go to memory only.
 */

const API_BASE_URL =
  process.env.REACT_APP_API_URL ||
  "https://gamespotbooking-v1-production.up.railway.app";

// ============================================================
// In-memory token store (NOT accessible to XSS attacks)
// ============================================================
let _accessToken = null;
let _isRefreshing = false;
let _refreshPromise = null;

/**
 * Set the access token in memory (called after login/refresh)
 */
export const setAccessToken = (token) => {
  _accessToken = token;
  // Also store in localStorage for backward compatibility with mobile
  // This will be removed in a future update once HttpOnly cookies are fully deployed
  try {
    if (token) {
      localStorage.setItem('gamespot_auth_token', token);
    } else {
      localStorage.removeItem('gamespot_auth_token');
    }
  } catch (e) { /* SSR or storage full */ }
};

/**
 * Get the current access token
 * Prefers in-memory token; falls back to localStorage for existing sessions
 */
export const getAuthToken = () => {
  if (_accessToken) return _accessToken;
  try {
    return localStorage.getItem('gamespot_auth_token');
  } catch (e) {
    return null;
  }
};

/**
 * Clear all auth tokens (called on logout)
 */
export const clearTokens = () => {
  _accessToken = null;
  try {
    localStorage.removeItem('gamespot_auth_token');
  } catch (e) { /* ignore */ }
};

/**
 * Attempt to refresh the access token using the HttpOnly refresh token cookie
 */
const refreshAccessToken = async () => {
  // Prevent multiple simultaneous refresh attempts
  if (_isRefreshing) {
    return _refreshPromise;
  }
  
  _isRefreshing = true;
  _refreshPromise = (async () => {
    try {
      const response = await fetch(`${API_BASE_URL}/api/auth/refresh`, {
        method: 'POST',
        credentials: 'include',  // Send the HttpOnly refresh cookie
        headers: { 'Content-Type': 'application/json' },
      });
      
      if (response.ok) {
        const data = await response.json();
        if (data.success && data.token) {
          setAccessToken(data.token);
          return true;
        }
      }
      return false;
    } catch (e) {
      return false;
    } finally {
      _isRefreshing = false;
      _refreshPromise = null;
    }
  })();
  
  return _refreshPromise;
};

export const apiFetch = async (path, options = {}, _isRetry = false) => {
  const url = `${API_BASE_URL}${path}`;
  
  // Get JWT token for authentication
  const token = getAuthToken();
  
  // Build headers with optional Authorization token
  const headers = {
    "Content-Type": "application/json",
    ...(options.headers || {}),
  };
  
  // Add Authorization header if we have a token
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }
  
  // Timeout for slow connections — 20s for login/auth, 30s for other requests
  // Forgot-password needs more time since it involves sending an email on the backend
  const isAuthPath = path.includes('/auth/');
  const isForgotPassword = path.includes('/forgot-password') || path.includes('/reset-password') || path.includes('/verify-reset-otp');
  const timeoutMs = isForgotPassword ? 45000 : isAuthPath ? 20000 : 30000;
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
  
  let response;
  try {
    response = await fetch(url, {
      ...options,
      credentials: "include",
      headers,
      signal: controller.signal,
    });
  } catch (err) {
    clearTimeout(timeoutId);
    if (err.name === 'AbortError') {
      throw new Error('Request timed out. Please check your internet connection and try again.');
    }
    throw new Error('Network error. Please check your internet connection.');
  }
  clearTimeout(timeoutId);

  const data = await response.json().catch(() => ({}));

  // Handle 401 Unauthorized
  if (response.status === 401) {
    // On first 401, try to refresh the token
    if (!_isRetry) {
      const refreshed = await refreshAccessToken();
      if (refreshed) {
        // Retry the original request with new token
        return apiFetch(path, options, true);
      }
    }
    
    // Refresh failed or this was already a retry — clear auth state
    clearTokens();
    
    // Redirect to login if on a protected page
    const currentPath = window.location.pathname;
    const publicPaths = ['/', '/login', '/signup', '/forgot-password', '/contact', '/games', '/updates', '/discount-game', '/win-free-game', '/instagram-promo', '/membership', '/get-offers'];
    if (!publicPaths.includes(currentPath) && data.redirect) {
      window.location.href = '/login';
      return data;
    }
  }

  // Handle 429 Too Many Requests - rate limited
  if (response.status === 429) {
    throw new Error(data.error || 'Too many requests. Please try again later.');
  }

  if (!response.ok) {
    throw new Error(data.error || "API request failed");
  }

  return data;
};

File: gamespot_app/lib/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../config/api_config.dart';

/// Central API client — mirrors the web app's apiClient.ts
/// Handles JWT auth, token refresh, retries, and error handling.
class ApiClient {
  static final ApiClient _instance = ApiClient._internal();
  factory ApiClient() => _instance;
  ApiClient._internal();

  final _storage = const FlutterSecureStorage();
  static const _tokenKey = 'gamespot_auth_token';
  static const _refreshKey = 'gamespot_refresh_token';

  String? _accessToken;

  /// Initialize — load token from secure storage
  Future<void> init() async {
    try {
      _accessToken = await _storage.read(key: _tokenKey);
    } catch (e) {
      // Keychain may not be available on simulator without entitlements
      _accessToken = null;
    }
  }

  /// Get current access token
  String? get accessToken => _accessToken;

  /// Set access token (after login)
  Future<void> setAccessToken(String? token) async {
    _accessToken = token;
    try {
      if (token != null) {
        await _storage.write(key: _tokenKey, value: token);
      } else {
        await _storage.delete(key: _tokenKey);
      }
    } catch (_) {
      // Keychain may not be available on simulator
    }
  }

  /// Clear all tokens (on logout)
  Future<void> clearTokens() async {
    _accessToken = null;
    try {
      await _storage.deleteAll();
    } catch (_) {
      // Keychain may not be available on simulator
    }
  }

  /// Build headers for requests
  Map<String, String> _headers({bool withContentType = true}) {
    final headers = <String, String>{};
    if (withContentType) {
      headers['Content-Type'] = 'application/json';
    }
    if (_accessToken != null) {
      headers['Authorization'] = 'Bearer $_accessToken';
    }
    return headers;
  }

  /// Refresh access token
  Future<bool> _refreshAccessToken() async {
    try {
      final response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}${ApiConfig.refreshToken}'),
        headers: {'Content-Type': 'application/json'},
      );
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true && data['token'] != null) {
          await setAccessToken(data['token']);
          return true;
        }
      }
      return false;
    } catch (e) {
      return false;
    }
  }

  /// Core fetch method with retry logic (mirrors fetchWithCredentials from web)
  Future<Map<String, dynamic>> fetch(
    String path, {
    String method = 'GET',
    Map<String, dynamic>? body,
    int retries = 3,
    bool isRetry = false,
  }) async {
    final url = Uri.parse('${ApiConfig.baseUrl}$path');
    final needsContentType = ['POST', 'PUT', 'PATCH', 'DELETE'].contains(method);

    http.Response response;
    try {
      switch (method) {
        case 'POST':
          response = await http
              .post(url,
                  headers: _headers(withContentType: needsContentType),
                  body: body != null ? json.encode(body) : null)
              .timeout(const Duration(seconds: 40));
          break;
        case 'PUT':
          response = await http
              .put(url,
                  headers: _headers(withContentType: needsContentType),
                  body: body != null ? json.encode(body) : null)
              .timeout(const Duration(seconds: 40));
          break;
        case 'DELETE':
          response = await http
              .delete(url, headers: _headers(withContentType: needsContentType))
              .timeout(const Duration(seconds: 40));
          break;
        default:
          response = await http
              .get(url, headers: _headers(withContentType: false))
              .timeout(const Duration(seconds: 40));
      }
    } catch (e) {
      if (retries > 0) {
        await Future.delayed(Duration(milliseconds: 2000 + (3 - retries) * 1500));
        return fetch(path, method: method, body: body, retries: retries - 1);
      }
      throw ApiException('Unable to connect to the server. Please try again.');
    }

    // Retry on 502/503/504 (Railway cold start)
    if ([502, 503, 504].contains(response.statusCode) && retries > 0) {
      await Future.delayed(const Duration(milliseconds: 2500));
      return fetch(path, method: method, body: body, retries: retries - 1);
    }

    Map<String, dynamic> data;
    try {
      data = json.decode(response.body);
    } catch (e) {
      if (retries > 0) {
        await Future.delayed(const Duration(milliseconds: 2000));
        return fetch(path, method: method, body: body, retries: retries - 1);
      }
      throw ApiException('Server returned an unexpected response. Please try again.');
    }

    // Handle 401 — try refresh
    if (response.statusCode == 401 && !isRetry) {
      final refreshed = await _refreshAccessToken();
      if (refreshed) {
        return fetch(path, method: method, body: body, isRetry: true);
      }
      await clearTokens();
    }

    if (response.statusCode == 429) {
      throw ApiException(data['error'] ?? 'Too many requests. Please try again later.');
    }

    if (response.statusCode >= 400) {
      throw ApiException(data['error'] ?? 'Something went wrong. Please try again.');
    }

    return data;
  }

  // Convenience methods
  Future<Map<String, dynamic>> get(String path) => fetch(path);

  Future<Map<String, dynamic>> post(String path, {Map<String, dynamic>? body}) =>
      fetch(path, method: 'POST', body: body);

  Future<Map<String, dynamic>> put(String path, {Map<String, dynamic>? body}) =>
      fetch(path, method: 'PUT', body: body);

  Future<Map<String, dynamic>> delete(String path) =>
      fetch(path, method: 'DELETE');
}

class ApiException implements Exception {
  final String message;
  ApiException(this.message);

  @override
  String toString() => message;
}
